internal/tasks/publishertask.go

```diff
@@ -0,0 +1,149 @@
+package tasks
+
+import (
+	"context"
+
+	"github.com/razorpay/metro/internal/common"
+	"github.com/razorpay/metro/internal/nodebinding"
+	"github.com/razorpay/metro/internal/topic"
+	"github.com/razorpay/metro/pkg/logger"
+	"github.com/razorpay/metro/pkg/registry"
+
+	"golang.org/x/sync/errgroup"
+)
+
+// PublisherTask implements the Watcher and maintains a pre-warmup.
+type PublisherTask struct {
+	id              string
+	registry        registry.IRegistry
+	topicCore       topic.ICore
+	nodeBindingCore nodebinding.ICore
+	topicWatchData  chan *struct{}
+	topicCacheData  map[string]bool
+}
+
+// NewPublisherTask creates PublisherTask instance
+func NewPublisherTask(
+	id string,
+	registry registry.IRegistry,
+	topicCore topic.ICore,
+	nodeBindingCore nodebinding.ICore,
+	options ...Option,
+) (IPubTask, error) {
+	publisherTask := &PublisherTask{
+		id:              id,
+		registry:        registry,
+		topicCore:       topicCore,
+		nodeBindingCore: nodeBindingCore,
+		topicWatchData:  make(chan *struct{}),
+		topicCacheData:  make(map[string]bool),
+	}
+
+	for _, option := range options {
+		option(publisherTask)
+	}
+
+	return publisherTask, nil
+}
+
+// Run the task
+func (pu *PublisherTask) Run(ctx context.Context) error {
+	var err error
+	var topicWatcher registry.IWatcher
+	err = pu.refreshCache(ctx)
+	if err != nil {
+		return err
+	}
+
+	twh := registry.WatchConfig{
+		WatchType: "keyprefix",
+		WatchPath: common.GetBasePrefix() + topic.Prefix,
+		Handler: func(ctx context.Context, pairs []registry.Pair) {
+			logger.Ctx(ctx).Infow("PublisherTask: topic watch handler data", "pairs", pairs)
+			pu.topicWatchData <- &struct{}{}
+		},
+	}
+
+	topicWatcher, err = pu.registry.Watch(ctx, &twh)
+	if err != nil {
+		return err
+	}
+
+	leadgrp, gctx := errgroup.WithContext(ctx)
+
+	// watch the Topic path for new topics and rebalance
+	leadgrp.Go(func() error {
+		watchErr := topicWatcher.StartWatch()
+		close(pu.topicWatchData)
+		return watchErr
+	})
+
+	// handle topic updates
+	leadgrp.Go(func() error {
+		for {
+			select {
+			case <-gctx.Done():
+				return gctx.Err()
+			case val := <-pu.topicWatchData:
+				if val == nil {
+					continue
+				}
+				terr := pu.refreshCache(ctx)
+				if terr != nil {
+					logger.Ctx(gctx).Errorw("PublisherTask: Failed to refresh cache for topic/subscription/nodes", "error", err.Error())
+				} else {
+					logger.Ctx(gctx).Infow("PublisherTask: Topic Cache refreshed")
+				}
+			}
+		}
+	})
+
+	// wait for done channel to be closed and stop watches if received done
+	leadgrp.Go(func() error {
```

Comment:
rather than introducing separate go routine to close topic watcher, can we implement in the above ctx done select case..?

Reply:
@sakshi-goyal-razorpay the purpose of this is to wait for the done channel to be closed and then stop watcher if received done as mentioned in the comment above.